onnxruntime/core/providers/cpu/math/einsum_utils/einsum_typed_compute_processor.cc

```diff
@@ -0,0 +1,359 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "einsum_typed_compute_processor.h"
+
+namespace onnxruntime {
+
+template <typename T>
+void EinsumTypedComputeProcessor<T>::FinalizeOutput(const Tensor& candidate_output,
+                                                    const std::vector<int64_t>& ordered_subscript_indices_in_candidate) {
+  const std::vector<int64_t>& subscript_indices_to_output_indices = einsum_compute_preprocessor_.GetMappedSubscriptIndicesToOutputindices();
+  const auto& output_dims = einsum_compute_preprocessor_.GetOutputDims();
+  TensorShape output_shape = TensorShape(output_dims);
+  const auto output_rank = output_dims.size();
+  Tensor& output = *context_->Output(0, output_dims);
+
+  ORT_ENFORCE(candidate_output.Shape().Size() == output_shape.Size(),
+              "Einsum op: The candidate output cannot be reshaped into the op's output");
+
+  const auto& candidate_output_dims = candidate_output.Shape().GetDims();
+  const auto candidate_output_rank = candidate_output_dims.size();
+
+  // This vector holds the shape of the candidate_output after removing the dims that have
+  // been reduced in the final output
+  std::vector<int64_t> candidate_output_shape_without_reduced_dims;
+  candidate_output_shape_without_reduced_dims.reserve(candidate_output_rank);  // reserve upper bound
+
+  // Identify the permutation required by the op's output
+  std::vector<size_t> output_permutation;
+  output_permutation.resize(output_rank, 0);
+  size_t output_iter = 0;
+
+  for (size_t iter = 0; iter < ordered_subscript_indices_in_candidate.size(); ++iter) {
+    auto output_index = subscript_indices_to_output_indices[ordered_subscript_indices_in_candidate[iter]];
+
+    // If output_index is -1, then this dimension does not show up in the op's output and has been reduced along the way
+    if (output_index != -1) {
+      output_permutation[output_index] = output_iter++;
+      candidate_output_shape_without_reduced_dims.push_back(candidate_output_dims[iter]);
+    } else {
+      // This dim doesn't show up in the op's output and hence we check if the dim has been reduced in the candidate output
+      ORT_ENFORCE(candidate_output_dims[iter] == 1, "Not all dimensions to be reduced have been reduced in the candidate output");
+    }
```

Comment:
nit: would be good to output candidate_output_dims. can leverage the ostream helper for TensorShape
      ORT_ENFORCE(candidate_output_dims[iter] == 1, "Not all dimensions to be reduced have been reduced in the candidate output. Candidate output dims:", candidate_output.Shape());

Reply:
Done